### PR TITLE
fix: [DEL-4269]: Properly retrying websocket connection if it fails

### PR DESCRIPTION
**Pull request by mdmilic**
## Describe your changes
Websocket connection wasn't properly being retried when it fails. We were essentially having only one retry attempt which  was waiting indefinitely to succeed, but in our case the request failed and the Future got canceled (but delegate continued waiting on it). 
Having timeout helps with this, but also we are now attempting reconnect until the connection state is not erroneous, as so far even if connection attempt fails we would stop trying. 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34849)

### CODE REVIEW

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 822 at r2](https://reviewable.io/reviews/harness/harness-core/34849#-N5lV8Mg6oClF0d1_tz0:-N5lV8Mg6oClF0d1_tz1:byrngo2) ([raw file](https://github.com/harness/harness-core/blob/05f578fa66d64a34a0b04e0f683aa00abb0c8d35/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L822)):*
> ```Java
>         trySocketReconnect();
>       } else if (e instanceof ConnectException) {
>         log.warn("Failed to connect.", e);
> ```

why do we need to restart in this case?

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 834 at r2](https://reviewable.io/reviews/harness/harness-core/34849#-N5lVHhcBw1h5W_LyaHs:-N5lVHhcBw1h5W_LyaHt:br0s811) ([raw file](https://github.com/harness/harness-core/blob/05f578fa66d64a34a0b04e0f683aa00abb0c8d35/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L834)):*
> ```Java
>           log.error("Failed closing the socket!", ex);
>         }
>         restartNeeded.set(true);
> ```

why do we need to restart?

**Review by raghuAtWings — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34849)* status: 0 of 1 approvals obtained.

File: 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java [hunk 1]
```diff
@@ -799,61 +806,63 @@ private void handleOpen(Object o) {
...
+      if (e instanceof SSLException || e instanceof TransportNotSupported) {
+        log.warn("Reopening connection to manager because of exception", e);
         trySocketReconnect();
-      } finally {
-        reconnectingSocket.set(false);
+      } else if (e instanceof ConnectException) {
+        log.warn("Failed to connect.", e);
+        restartNeeded.set(true);
+      } else if (e instanceof ConcurrentModificationException) {
+        log.warn("ConcurrentModificationException on WebSocket ignoring");
```
Comment (bharatgoelharness) [inline 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java:825]: why do we need to levels of logging here

**Review by mdmilic — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34849)* status: 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 822 at r2](https://reviewable.io/reviews/harness/harness-core/34849#-N5lV8Mg6oClF0d1_tz0:-N5tkl2jE5p4Ykghnvpq:b6f1rf5) ([raw file](https://github.com/harness/harness-core/blob/05f578fa66d64a34a0b04e0f683aa00abb0c8d35/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L822)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

why do we need to restart in this case?
</blockquote></details>

This code is quite old, so I was trying to make the smallest change possible. By looking at old diffs, the logic why this got implemented in the first place doesn't exist any more. The logic was, delegate retries to open websocket connection 50 times, if that fails it requests to restart the entire thing. This exception could occur if manager is not reachable. With current code, if it's initial connect, delegate would just be stuck. If it got disconnected, it would just restart
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 825 at r2](https://reviewable.io/reviews/harness/harness-core/34849#-N5lLPcMzCpfjItjzmbr-r2-825:-N5tnj4g1UbM30nStq4n:btupv1y) ([raw file](https://github.com/harness/harness-core/blob/05f578fa66d64a34a0b04e0f683aa00abb0c8d35/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L825)):*
<details><summary><i>Previously, bharatgoelharness (Bharat Goel) wrote…</i></summary><blockquote>

why do we need to levels of logging here
</blockquote></details>

:man_shrugging: 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 834 at r2](https://reviewable.io/reviews/harness/harness-core/34849#-N5lVHhcBw1h5W_LyaHs:-N5tnzaI-48Rre-RSoWn:bks04da) ([raw file](https://github.com/harness/harness-core/blob/05f578fa66d64a34a0b04e0f683aa00abb0c8d35/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L834)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

why do we need to restart?
</blockquote></details>

I think similar as the other one. Some of the initial delegate code, I guess we didn't really know what to do so we requested restart. 
I didn't really change any of the existing exception handling just enhanced the retry and added some new handling for the specific exceptions that happen for this issue. This method only has indentation changes and I removed the restart lock as I put it in the tryRestart method

**Comment by raghuAtWings** [comment]
trigger messagecheck

**Review by raghuAtWings — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34849)* status: Some of the checks are too old. 1. unresolved discussions.

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 825 at r2](https://reviewable.io/reviews/harness/harness-core/34849#-N5lLPcMzCpfjItjzmbr-r2-825:-N6EwpWIEX7-ojhMabFw:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/05f578fa66d64a34a0b04e0f683aa00abb0c8d35/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L825)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

:man_shrugging: 
</blockquote></details>

done

**Review by raghuAtWings — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34849#-:-N6EwwNZBzjLlQLmU1n0:bnfp4nl)* status: Some of the checks are too old. 1. unresolved discussions.

**Review by bharatgoelharness — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34849)* status: Some of the checks are too old.

**Review by bharatgoelharness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34849#-:-N6GrhAQ3IIo2_8BFLkQ:bnfp4nl)* status: Some of the checks are too old.

**Pull request merged**